### ISSUE TITLE
[doc] view appears to mean different things, `view/reshape` vs `transpose/permute`.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#75510]
### 📚 The doc issue

From the docs it might appear that "view" and "shares the same storage" are interchangeable concepts. Unfortunately, this is not the case and a bit confusing.

For example, the doc for `transpose` is clear in saying that the returned tensor shares the same storage with the original input.
However, `permute` reads `Returns a view of the original tensor input with its dimensions permuted`. If we check the doc for `reshape`, it reads `When possible, the returned tensor will be a view of input. Otherwise, it will be a copy`, and references the `view` operation. The doc for `view`  defines when a tensor could be "viewed", and provides a relation between strides and shapes. Note, however, that `permute` is stride-agnostic. So there we have our first contradiction, when "view", `view` and "shares the same storage are not equivalent" when restricted to at least `view` and `permute`.

To illustrate, let's have a look at some examples:
```python
>>> import torch
>>> x = torch.rand(4, 3, 2) # contiguous
>>> y = x.reshape(2, 3, 4) # same storage
>>> z = x.permute((2, 1, 0)) # same storage
>>> x[0, 0, 0] = -100
>>> y[0, 0, 0]
tensor(-100.)
>>> z[0, 0, 0]
tensor(-100.)
>>> x = torch.rand(4, 2, 3).transpose(-1, -2) # NON-contiguous!!!
>>> y = x.reshape(2, 3, 4) # same storage???
>>> z = x.permute((2, 1, 0)) # same storage
>>> x[0, 0, 0] = -100
>>> y[0, 0, 0]
tensor(0.2377)
>>> z[0, 0, 0]
tensor(-100.)
```

We can see that `reshape` handles permutations of dimensions in the contiguous case, but makes a copy in the non-contiguous.
The user, if interested in memory efficiency, now has to differentiate between contiguous/non-contiguous cases and depending on that use either `permute` or `reshape`.

### Suggest a potential alternative/fix

"Shares the same storage" and a "view" are not the same in some cases with non-contiguous inputs.

Also, the documentation of `view` is not informative. It is really hard to grasp the underlying contiguity-like condition, as it is not clear what `d`, `k` even are and where they come from.

I think there is a much easier description of when `view/reshape` can operate over the input's storage, for example:
```
1. view/reshape will use the same storage if the input is contiguous,
    no matter what the shape is (as long as numel is preserved).

2. view/reshape will use the same storage if the input is non-contiguous,
    and its shape (d_1, ..., d_k) is reshaped into
    (d_{11}, ... d_{1 l_1}, ..., d_{k1}, ..., d_{k l_k}) as long as
     d_i = prod(d_{i1}, ..., d_{i l_i}) for each i in {1, ..., k}.

3. view/reshape will use the same storage if any 1-size dims
     are added to and/or removed from the original tensor.
```

Some examples testing this description.

Non-contiguous case:
```python
```

**Comment by ezyang** [comment]
Feel free to send some doc PRs.

I think trying to scrub confusion between "view" the concept and `view` the function will be difficult, however, because we can't rename the view function, and "is view of" is a more refined statement than "shares same storage": (1) two tensors can share storage and not alias each other, e.g., x[0] and x[1], (2) view is an autograd concept and there is a way to get aliasing tensors that aren't considered views as far as autograd is concerned.

**Comment by ngimel** [comment]
>So there we have our first contradiction, when "view", view and "shares the same storage are not equivalent" when restricted to at least view and permute.

Sorry, I don't see where's a contradiction here. "view" means using the same storage, "permute" is always a view, semantics of "permute" and "reshape" operations is different. 
Defining when exactly a tensor can be viewed is not trivial, a detailed description of "contiguous subspaces" is needed (and we are not even implementing it exactly right anyway), but current docs are doing a decent job.

**Comment by nikitaved** [comment]
Right, but `reshape` on a contiguous tensor is always a view, even if we reshape with perturbed dimensions, this does not hold for non-contiguous tensors.
